File: internal/llm/ollama.go
package llm

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
)

const (
	DefaultOllamaHost = "http://localhost:11434"
	defaultNumCtx     = 16384
)

type Ollama struct {
	host       string
	model      string
	numCtx     int
	httpClient *http.Client
}

func NewOllama(host, model string) *Ollama {
	if host == "" {
		host = DefaultOllamaHost
	}
	numCtx := defaultNumCtx
	if raw := os.Getenv("NINA_OLLAMA_NUM_CTX"); raw != "" {
		if parsed, err := strconv.Atoi(raw); err == nil && parsed > 0 {
			numCtx = parsed
		}
	}
	return &Ollama{host: host, model: model, numCtx: numCtx, httpClient: http.DefaultClient}
}

type ollamaMessage struct {
	Role      string           `json:"role"`
	Content   string           `json:"content"`
	ToolName  string           `json:"tool_name,omitempty"`
	ToolCalls []ollamaToolCall `json:"tool_calls,omitempty"`
}

type ollamaToolCall struct {
	Function ollamaFunctionCall `json:"function"`
}

type ollamaFunctionCall struct {
	Name      string          `json:"name"`
	Arguments json.RawMessage `json:"arguments"`
}

type ollamaTool struct {
	Type     string             `json:"type"`
	Function ollamaToolFunction `json:"function"`
}

type ollamaToolFunction struct {
	Name        string         `json:"name"`
	Description string         `json:"description"`
	Parameters  map[string]any `json:"parameters"`
}

type ollamaRequest struct {
	Model    string          `json:"model"`
	Messages []ollamaMessage `json:"messages"`
	Tools    []ollamaTool    `json:"tools,omitempty"`
	Stream   bool            `json:"stream"`
	Options  map[string]any  `json:"options,omitempty"`
}

type ollamaChunk struct {
	Message struct {
		Content   string           `json:"content"`
		ToolCalls []ollamaToolCall `json:"tool_calls"`
	} `json:"message"`
	Done       bool   `json:"done"`
	DoneReason string `json:"done_reason"`
	Error      string `json:"error"`
}

func (o *Ollama) Converse(ctx context.Context, conv *Conversation, onDelta func(string)) (Turn, error) {
	payload := ollamaRequest{
		Model:    o.model,
		Messages: ollamaMessages(conv),
		Tools:    ollamaTools(),
		Stream:   true,
		Options:  map[string]any{"num_ctx": o.numCtx},
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return Turn{}, err
	}
	request, err := http.NewRequestWithContext(ctx, http.MethodPost, o.host+"/api/chat", bytes.NewReader(body))
	if err != nil {
		return Turn{}, err
	}
	request.Header.Set("Content-Type", "application/json")

	response, err := o.httpClient.Do(request)
	if err != nil {
		return Turn{}, fmt.Errorf("calling ollama at %s: %w (is `ollama serve` running?)", o.host, err)
	}
	defer response.Body.Close()
	if response.StatusCode != http.StatusOK {
		return Turn{}, ollamaError(response)
	}

	turn := Turn{}
	callIndex := 0
	decoder := json.NewDecoder(response.Body)
	for {
		var chunk ollamaChunk
		if err := decoder.Decode(&chunk); err == io.EOF {
			break
		} else if err != nil {
			return Turn{}, fmt.Errorf("reading ollama stream: %w", err)
		}
		if chunk.Error != "" {
			return Turn{}, fmt.Errorf("ollama: %s", chunk.Error)
		}
		if chunk.Message.Content != "" {
			turn.Text += chunk.Message.Content
			if onDelta != nil {
				onDelta(chunk.Message.Content)
			}
		}
		for _, call := range chunk.Message.ToolCalls {
			turn.ToolCalls = append(turn.ToolCalls, ToolCall{
				ID:    fmt.Sprintf("call_%d_%d", len(conv.Messages), callIndex),
				Name:  call.Function.Name,
				Input: call.Function.Arguments,
			})
			callIndex++
		}
		if chunk.Done {
			turn.StopReason = chunk.DoneReason
			break
		}
	}
	if len(turn.ToolCalls) > 0 {
		turn.StopReason = "tool_use"
	}
	conv.addAssistantTurn(turn)
	return turn, nil
}

func ollamaError(response *http.Response) error {
	raw, _ := io.ReadAll(io.LimitReader(response.Body, 4096))
	var body struct {
		Error string `json:"error"`
	}
	if json.Unmarshal(raw, &body) == nil && body.Error != "" {
		return fmt.Errorf("ollama: %s", body.Error)
	}
	return fmt.Errorf("ollama: unexpected status %s: %s", response.Status, raw)
}

func ollamaMessages(conv *Conversation) []ollamaMessage {
	out := make([]ollamaMessage, 0, len(conv.Messages)+1)
	if conv.System != "" {
		out = append(out, ollamaMessage{Role: "system", Content: conv.System})
	}
	for _, message := range conv.Messages {
		if len(message.ToolResults) > 0 {
			for _, result := range message.ToolResults {
				content := result.Content
				if result.IsError {
					content = "Error: " + content
				}
				out = append(out, ollamaMessage{
					Role:     "tool",
					Content:  content,
					ToolName: conv.toolNameFor(result.ToolCallID),
				})
			}
			continue
		}
		converted := ollamaMessage{Role: message.Role, Content: message.Text}
		for _, call := range message.ToolCalls {
			converted.ToolCalls = append(converted.ToolCalls, ollamaToolCall{
				Function: ollamaFunctionCall{Name: call.Name, Arguments: call.Input},
			})
		}
		out = append(out, converted)
	}
	return out
}

func ollamaTools() []ollamaTool {
	defs := ToolDefs()
	out := make([]ollamaTool, 0, len(defs))
	for _, def := range defs {
		out = append(out, ollamaTool{
			Type: "function",
			Function: ollamaToolFunction{
				Name:        def.Name,
				Description: def.Description,
				Parameters: map[string]any{
					"type":       "object",
					"properties": def.Properties,
					"required":   def.Required,
				},
			},
		})
	}
	return out
}

File: src/nina/agent/ollama.py
from __future__ import annotations

import contextlib
import json
import os
from collections.abc import AsyncIterator

import httpx

from nina.agent.session import AgentEvent, TextDelta, ToolHandler, TurnComplete
from nina.agent.tools import TOOL_SPECS, TOOL_SUBMIT_REVIEW
from nina.system import state

DEFAULT_HOST = "http://localhost:11434"
DEFAULT_NUM_CTX = 16384
MAX_TURNS = 40


def _resolve_host(host: str | None) -> str:
    return host or os.environ.get("NINA_OLLAMA_HOST") or DEFAULT_HOST


def _resolve_num_ctx(num_ctx: int | None) -> int:
    if num_ctx is not None:
        return num_ctx
    raw = os.environ.get("NINA_OLLAMA_NUM_CTX", "")
    if raw:
        with contextlib.suppress(ValueError):
            parsed = int(raw)
            if parsed > 0:
                return parsed
    return DEFAULT_NUM_CTX


def _connect_error(host: str, err: Exception) -> RuntimeError:
    return RuntimeError(f"calling ollama at {host}: {err} (is `ollama serve` running?)")


async def _status_error(response: httpx.Response) -> RuntimeError:
    raw = await response.aread()
    with contextlib.suppress(json.JSONDecodeError):
        body = json.loads(raw)
        message = body.get("error")
        if message:
            return RuntimeError(f"ollama: {message}")
    text = raw.decode(errors="replace")
    return RuntimeError(f"ollama: unexpected status {response.status_code}: {text}")


def _ollama_tools() -> list[dict[str, object]]:
    return [
        {
            "type": "function",
            "function": {
                "name": spec.name,
                "description": spec.description,
                "parameters": spec.schema,
            },
        }
        for spec in TOOL_SPECS
    ]


async def classify(
    system_prompt: str,
    prompt: str,
    model: str,
    host: str | None = None,
    transport: httpx.AsyncBaseTransport | None = None,
) -> str:
    resolved_host = _resolve_host(host)
    payload = {
        "model": model,
        "messages": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": prompt},
        ],
        "stream": False,
    }
    async with httpx.AsyncClient(timeout=None, transport=transport) as client:
        try:
            response = await client.post(f"{resolved_host}/api/chat", json=payload)
        except httpx.HTTPError as err:
            raise _connect_error(resolved_host, err) from err
        if response.status_code != 200:
            raise await _status_error(response)
        body = response.json()
    error = body.get("error")
    if error:
        raise RuntimeError(f"ollama: {error}")
    return str(body.get("message", {}).get("content", ""))


class OllamaAgentSession:
    def __init__(
        self,
        system_prompt: str,
        cwd: str,
        handlers: dict[str, ToolHandler],
        model: str,
        host: str | None = None,
        num_ctx: int | None = None,
        history: list[dict[str, object]] | None = None,
        transport: httpx.AsyncBaseTransport | None = None,
    ) -> None:
        self._cwd = cwd
        self._handlers = handlers
        self._model = model
        self._host = _resolve_host(host)
        self._num_ctx = _resolve_num_ctx(num_ctx)
        self._transport = transport
        self._messages: list[dict[str, object]] = [{"role": "system", "content": system_prompt}]
        for entry in history or []:
            role, text = entry.get("role"), entry.get("text")
            if role in ("user", "assistant") and text:
                self._messages.append({"role": role, "content": text})
        self.session_id: str | None = None

    async def send(self, text: str) -> AsyncIterator[AgentEvent]:
        state.append_transcript_safe(self._cwd, {"role": "user", "text": text})
        self._messages.append({"role": "user", "content": text})
        async with httpx.AsyncClient(timeout=None, transport=self._transport) as client:
            for _ in range(MAX_TURNS):
                assistant_text = ""
                tool_calls: list[dict[str, object]] = []
                async for event in self._stream_turn(client):
                    if isinstance(event, TextDelta):
                        assistant_text += event.text
                        yield event
                    else:
                        tool_calls.append(event)
                self._messages.append(
                    {
                        "role": "assistant",
                        "content": assistant_text,
                        **({"tool_calls": tool_calls} if tool_calls else {}),
                    }
                )
                if assistant_text:
                    state.append_transcript_safe(
                        self._cwd,
                        {"role": "assistant", "text": assistant_text, "model": self._model},
                    )
                if not tool_calls:
                    break
                reviewed = await self._run_tool_calls(tool_calls)
                if reviewed:
                    break
        yield TurnComplete()

    async def _stream_turn(
        self, client: httpx.AsyncClient
    ) -> AsyncIterator[TextDelta | dict[str, object]]:
        payload = {
            "model": self._model,
            "messages": self._messages,
            "tools": _ollama_tools(),
            "stream": True,
            "options": {"num_ctx": self._num_ctx},
        }
        try:
            async with client.stream("POST", f"{self._host}/api/chat", json=payload) as response:
                if response.status_code != 200:
                    raise await _status_error(response)
                async for line in response.aiter_lines():
                    if not line.strip():
                        continue
                    chunk = json.loads(line)
                    if chunk.get("error"):
                        raise RuntimeError(f"ollama: {chunk['error']}")
                    message = chunk.get("message", {})
                    piece = message.get("content", "")
                    if piece:
                        yield TextDelta(piece)
                    for call in message.get("tool_calls", []) or []:
                        yield call
                    if chunk.get("done"):
                        break
        except httpx.HTTPError as err:
            raise _connect_error(self._host, err) from err

    async def _run_tool_calls(self, tool_calls: list[dict[str, object]]) -> bool:
        reviewed = False
        for call in tool_calls:
            function = call.get("function", {})
            if not isinstance(function, dict):
                function = {}
            name = str(function.get("name", ""))
            raw_args = function.get("arguments", {})
            args = raw_args if isinstance(raw_args, dict) else json.loads(raw_args or "{}")
            handler = self._handlers.get(name)
            if handler is None:
                content, is_error = f"unknown tool {name!r}", True
            else:
                result = await handler(args)
                content, is_error = result.content, result.is_error
            self._messages.append(
                {
                    "role": "tool",
                    "tool_name": name,
                    "content": f"Error: {content}" if is_error else content,
                }
            )
            if name == TOOL_SUBMIT_REVIEW:
                reviewed = True
        return reviewed

    async def close(self) -> None:
        return None
